### PR TITLE
Unnest dictionary blocks on compaction

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: presto-spi/src/main/java/io/prestosql/spi/block/DictionaryBlock.java [hunk 1]
```diff
@@ -93,6 +93,9 @@ public DictionaryBlock(int idsOffset, int positionCount, Block dictionary, int[]
         this.retainedSizeInBytes = INSTANCE_SIZE + dictionary.getRetainedSizeInBytes() + sizeOf(ids);
 
         if (dictionaryIsCompacted) {
+            if (dictionary instanceof DictionaryBlock) {
+                throw new IllegalArgumentException("compacted dictionary should not have dictionary base block");
```
Comment (findepi) [inline presto-spi/src/main/java/io/prestosql/spi/block/DictionaryBlock.java]: `checkArgument(!(dictionary instanceof DictionaryBlock), ...`

File: presto-spi/src/main/java/io/prestosql/spi/block/DictionaryBlock.java [hunk 2]
```diff
@@ -464,4 +481,27 @@ public DictionaryBlock compact()
...
     }
+
+    private DictionaryBlock unnest()
+    {
+        if (!(dictionary instanceof DictionaryBlock)) {
+            return this;
+        }
+
+        int[] ids = new int[positionCount];
+        for (int i = 0; i < positionCount; ++i) {
```
Comment (findepi) [inline presto-spi/src/main/java/io/prestosql/spi/block/DictionaryBlock.java]: `i++`

File: presto-spi/src/main/java/io/prestosql/spi/block/DictionaryBlock.java [hunk 3]
```diff
@@ -464,4 +481,27 @@ public DictionaryBlock compact()
...
+            ids[i] = getId(i);
+        }
+
+        Block dictionary = this.dictionary;
+        while (dictionary instanceof DictionaryBlock) {
+            DictionaryBlock nestedDictionary = (DictionaryBlock) dictionary;
+            for (int i = 0; i < positionCount; ++i) {
```
Comment (findepi) [inline presto-spi/src/main/java/io/prestosql/spi/block/DictionaryBlock.java]: `i++`

File: presto-spi/src/main/java/io/prestosql/spi/block/DictionaryBlock.java (hunk 1 above)
Comment (sopel39) [reply presto-spi/src/main/java/io/prestosql/spi/block/DictionaryBlock.java]: No guava in SPI

File: presto-main/src/test/java/io/prestosql/block/TestDictionaryBlock.java [hunk 4]
```diff
@@ -156,6 +156,23 @@ public void testCompact()
...
+        DictionaryBlock nestedDictionary = new DictionaryBlock(createSlicesBlock(expectedValues), new int[] {0, 1, 2, 2, 4, 5});
+        DictionaryBlock dictionary = new DictionaryBlock(nestedDictionary, new int[] {2, 3, 2, 0});
+
+        assertEquals(
+                dictionary.getSizeInBytes(),
+                nestedDictionary.getPositionsSizeInBytes(new boolean[] {true, false, true, true, false, false}) + 4 * Integer.BYTES);
+        assertFalse(dictionary.isCompact());
+
+        DictionaryBlock compactBlock = dictionary.compact();
+        assertBlock(compactBlock.getDictionary(), TestDictionaryBlock::createBlockBuilder, new Slice[] {expectedValues[2], expectedValues[0]});
```
Comment (dain) [inline presto-main/src/test/java/io/prestosql/block/TestDictionaryBlock.java]: maybe add `assertInstanceOf(compactBlock.getDictionary(), VariableWidthBlock.class)` to make is super clear to future readers.

**Pull request merged**